src/descriptors/element_clip_edge.js

```diff
@@ -0,0 +1,195 @@
+var ElementSize = require("./element_size.js");
+
+var TOP = "top";
+var RIGHT = "right";
+var BOTTOM = "bottom";
+var LEFT = "left";
+
+function ElementClipEdge(element, position) {
+	var QElement = require("../q_element.js");      // break circular dependency
+	ensure.signature(arguments, [ QElement, String ]);
+
+	if (position === LEFT || position === RIGHT) PositionDescriptor.x(this);
+	else if (position === TOP || position === BOTTOM) PositionDescriptor.y(this);
+	else ensure.unreachable("Unknown position: " + position);
+
+	this._element = element;
+	this._position = position;
+}
+
+PositionDescriptor.extend(ElementClipEdge);
+
+ElementClipEdge.top = factoryFn(TOP);
+ElementClipEdge.right = factoryFn(RIGHT);
+ElementClipEdge.bottom = factoryFn(BOTTOM);
+ElementClipEdge.left = factoryFn(LEFT);
+
+ElementClipEdge.prototype.value = function value() {
+	ensure.signature(arguments, []);
+
+	var clipPosition = this.getRawClipPosition();
+
+	if (!clipPosition) {
+		throw new ClipNotAppliedException(ElementClipEdge.prototype.value,
+			"clip " + this._position + " css style not applied to " + this._element);
+	}
+
+	var edge = clipPosition[this._position];
+	var scroll = this._element.frame.getRawScrollPosition();
+
+	if (this._position === RIGHT || this._position === LEFT) return Position.x(edge + scroll.x);
+	else return Position.y(edge + scroll.y);
+};
+
+ElementClipEdge.prototype.toString = function toString() {
+	ensure.signature(arguments, []);
+	return "clip " + this._position + " edge of " + this._element;
+};
+
+function factoryFn(position) {
+	return function factory(element) {
+		return new ElementClipEdge(element, position);
+	};
+}
+
+var UNSET_CLIP_STYLES = [ "", "auto", "unset", "initial" ];
+var CLIP_RECT_PATTERN = /rect[\s]*\([\s]*([^\s,]+)[\s,]+([^\s,]+)[\s,]+([^\s,]+)[\s,]+([^\s,]+)[\s]*\)/;
+
+ElementClipEdge.prototype.getRawClipPosition = function getRawClipPosition() {
+	ensure.signature(arguments, []);
+
+	var rect = this._element.getRawStyle("clip");
+
+	if (rect === "") {
+		// As a fallback for IE8 for when it can't fork over the original clip css style, try generating a clip rect
+		// using clip components that currentStyle may have.  If we see non-empty strings for all four components, we'll
+		// build out a clip rect expression here ...
+
+		var clipLeft = this._element.getRawStyle("clip-left");
+		var clipRight = this._element.getRawStyle("clip-right");
+		var clipBottom = this._element.getRawStyle("clip-bottom");
+		var clipTop = this._element.getRawStyle("clip-top");
+
+		if (clipLeft && clipRight && clipBottom && clipTop) {
+			rect = "rect(" + clipTop + " " + clipRight + " " + clipBottom + " " + clipLeft + ")";
+		}
+	}
+
+	for (var i = 0, ii = UNSET_CLIP_STYLES.length; i < ii; i++) {
+		if (UNSET_CLIP_STYLES[i] === rect) {
+			return null;
+		}
+	}
+
+	var matches = rect.match(CLIP_RECT_PATTERN);
+	if (!matches) {
+		ensure.unreachable("Unknown clip css style: " + rect);
+	}
```

Comment:
Style: For single-line branches, particularly guard clauses like this one, don't use braces:

```javascript
if (!matches) ensure.unreachable("Unknown clip css style: " + rect);
```